src/compton2/Compton2.cc

```diff
@@ -0,0 +1,475 @@
+                              num_evals_,        num_points_, data_size};
+  rtt_c4::broadcast(&pack[0], pack.size(), bcast_rank);
+  UINT p = 0;
+  num_temperatures_ = pack[p++];
+  num_groups_ = pack[p++];
+  num_leg_moments_ = pack[p++];
+  num_evals_ = pack[p++];
+  num_points_ = pack[p++];
+  data_size = pack[p++];
+
+  // Derived sizes
+  const auto tsz = int(num_temperatures_);
+  const auto egsz = int(num_groups_ + 1U);
+  const auto fgsz = int(num_temperatures_ * num_groups_);
+  const auto isz = int(fgsz + 1U);
+  const auto dsz = int(data_size);
+
+  // Broadcast grids
+  if (rank != bcast_rank)
+    Ts_.resize(tsz);
+  rtt_c4::broadcast(&Ts_[0], tsz, 0);
+  if (rank != bcast_rank)
+    Egs_.resize(egsz);
+  rtt_c4::broadcast(&Egs_[0], egsz, 0);
+
+  // Broadcast sparse data structures
+  if (rank != bcast_rank)
+    first_groups_.resize(fgsz);
+  rtt_c4::broadcast(&first_groups_[0], fgsz, 0);
+  if (rank != bcast_rank)
+    indexes_.resize(isz);
+  rtt_c4::broadcast(&indexes_[0], isz, 0);
+
+  // Broadcast data itself
+  if (rank != bcast_rank)
+    data_.resize(dsz);
+  rtt_c4::broadcast(&data_[0], dsz, 0);
+  if (rank != bcast_rank)
+    derivs_.resize(dsz);
+  rtt_c4::broadcast(&derivs_[0], dsz, 0);
+}
+
+int Compton2::read_binary(std::string filename) {
+  // Read
+  auto fin = std::ifstream(filename, std::ios::in | std::ios::binary);
+
+  // Ensure valid type
+  char expected[] = " csk ";
+  char file_type[sizeof(expected)];
+  fin.read(file_type, sizeof(file_type));
+  if (std::strcmp(file_type, expected) != 0) {
+    std::cerr << "Expecting binary file " << filename << " to start with '" << expected
+              << "' but got '" << file_type << "'";
+    std::cerr << std::endl;
+    return 1;
+  }
+
+  UINT binary_ordering, version_major, version_minor;
+  fin.read(reinterpret_cast<char *>(&version_major), sizeof(UINT));
+  fin.read(reinterpret_cast<char *>(&version_minor), sizeof(UINT));
+  fin.read(reinterpret_cast<char *>(&binary_ordering), sizeof(UINT));
+  if (version_major != 1 || binary_ordering > 1) {
+    std::cerr << "Expecting a CSK binary file (version 1) with ordering 0 or 1 "
+                 "but got "
+              << version_major << " with ordering " << binary_ordering;
+    std::cerr << std::endl;
+    return 2;
+  }
+
+  constexpr UINT n = 7;
+  std::array<UINT, n> szs;
+  for (UINT i = 0; i < n; ++i)
+    fin.read(reinterpret_cast<char *>(&szs[i]), sizeof(szs[i]));
+  UINT j = 0;
+  UINT tsz = szs[j++];
+  UINT gsz = szs[j++];
+  UINT lsz = szs[j++];
+  UINT esz = szs[j++];
+  UINT fgsz = szs[j++];
+  UINT isz = szs[j++];
+  UINT dsz = szs[j++];
+
+  num_temperatures_ = tsz;
+  num_groups_ = gsz;
+  num_leg_moments_ = lsz;
+  num_evals_ = esz;
+  // point is (leg moment, eval) pair
+  // first eval has all leg moments; others have only the 0th moment
+  num_points_ = num_evals_ + num_leg_moments_ - 1U;
+  UINT egsz = gsz + 1;
+
+  if (false) {
```

Comment:
Remove or add a comment.  Consider `#if 0` to keep this out of the binary.

Reply:
Yeah, this is a debug print statement. Do you think there's value in making a function that prints the contents of a file or should I just remove this (or keep around with `#if 0`)?

Reply:
I really think that decision is up to you.  In more mature code, I would recommend that the code be removed entirely. Since this is brand new code, it might be worth keeping it in place locally.  Optionally, you could create a utility application that reads the file and dumps the contents to the screen.  We have something similar for both cdi_ipcress and cdi_eospac.